flink-ml-lib/src/main/java/org/apache/flink/ml/anomalydetection/isolationforest/IsolationForest.java

```diff
@@ -0,0 +1,611 @@
+            samplesData.add(streamRecord.getValue());
+        }
+
+        @Override
+        public void processElement2(StreamRecord<IForest> streamRecord) throws Exception {
+            Preconditions.checkState(!samplesDataCenter.get().iterator().hasNext());
+            samplesDataCenter.add(streamRecord.getValue());
+        }
+
+        @Override
+        public void onEpochWatermarkIncremented(
+                int epochWatermark, Context context, Collector<IForest> collector)
+                throws Exception {
+            IForest centers =
+                    Objects.requireNonNull(
+                            OperatorStateUtils.getUniqueElement(samplesDataCenter, "centers")
+                                    .orElse(null));
+            Iterator<DenseVector[]> samplesDataIterator = samplesData.get().iterator();
+            List<DenseVector> list = new ArrayList<>();
+            while (samplesDataIterator.hasNext()) {
+                DenseVector[] sampleData = samplesDataIterator.next();
+                list.add(centers.calculateAnomalyScore(sampleData));
+            }
+            DenseVector[] scores = list.toArray(new DenseVector[0]);
+            samplesDataScores.add(scores);
+
+            collector.collect(samplesDataCenter.get().iterator().next());
+            samplesDataCenter.clear();
+        }
+
+        @Override
+        public void onIterationTerminated(Context context, Collector<IForest> collector)
+                throws Exception {
+            IForest centers =
+                    Objects.requireNonNull(
+                            OperatorStateUtils.getUniqueElement(samplesDataCenter, "centers")
+                                    .orElse(null));
+            double centers0Sum1 = 0.0, centers1Sum1 = 0.0, centers0Sum2 = 0.0, centers1Sum2 = 0.0;
+            int size1 = 0, size2 = 0;
+            Iterator<DenseVector[]> samplesDataScoresIterator = samplesDataScores.get().iterator();
+            while (samplesDataScoresIterator.hasNext()) {
+                for (DenseVector denseVector : samplesDataScoresIterator.next()) {
+                    DenseVector denseVector1 = centers.classifyByCluster(denseVector);
+                    centers0Sum1 += denseVector1.get(0);
+                    centers1Sum1 += denseVector1.get(1);
+                    size1++;
+                }
+                centers0Sum2 = centers0Sum1 / size1;
+                centers1Sum2 = centers1Sum1 / size1;
+                size2++;
+            }
+
+            centers.center0 = centers0Sum2 / size2;
+            centers.center1 = centers1Sum2 / size2;
+
+            context.output(modelDataOutputTag, centers);
+
+            samplesDataCenter.clear();
+            samplesDataScores.clear();
+            samplesData.clear();
+        }
+    }
+
+    /** Construct isolation forest. */
+    public static class IForest implements Serializable {
+        public int treesNumber;
+        public int iters;
```

Comment:
Given that `IForest` is used as the model data, it seems a bit weird to have `iters` in the model data.

Reply:
The consideration in this aspect is that the follow-up needs to rely on IForest to calculate the distance from center0 and center1, so as to judge which point is closer to this point. IForest and the maximum number of averages retained by ITreeList need to be used.

Reply:
Only used in calculating center0 and center1, I will rethink it.